pkg/chunk/cache/redis_client.go

```diff
@@ -0,0 +1,303 @@
+		Addr:        cfg.Endpoint,
+		Password:    cfg.Password.Value,
+		PoolSize:    cfg.PoolSize,
+		IdleTimeout: cfg.IdleTimeout,
+		MaxConnAge:  cfg.MaxConnAge,
+	}
+	if cfg.EnableTLS {
+		opt.TLSConfig = &tls.Config{}
+	}
+	return &redisBasicClient{
+		logger:     logger,
+		expiration: cfg.Expiration,
+		timeout:    cfg.Timeout,
+		rdb:        redis.NewClient(opt),
+	}
+}
+
+func newRedisClusterClient(cfg *RedisConfig, logger log.Logger) *redisBasicClient {
+	opt := &redis.ClusterOptions{
+		Addrs:       strings.Split(cfg.Endpoint, ","),
+		Password:    cfg.Password.Value,
+		PoolSize:    cfg.PoolSize,
+		IdleTimeout: cfg.IdleTimeout,
+		MaxConnAge:  cfg.MaxConnAge,
+	}
+	if cfg.EnableTLS {
+		opt.TLSConfig = &tls.Config{}
+	}
+	return &redisBasicClient{
+		logger:     logger,
+		expiration: cfg.Expiration,
+		timeout:    cfg.Timeout,
+		rdb:        redis.NewClusterClient(opt),
+	}
+}
+
+func (c *redisBasicClient) Ping(ctx context.Context) error {
+	var cancel context.CancelFunc
+	if c.timeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, c.timeout)
+		defer cancel()
+	}
+
+	pong, err := c.rdb.Ping(ctx).Result()
+	if err != nil {
+		return err
+	}
+	if pong != "PONG" {
+		return fmt.Errorf("redis: Unexpected PING response %q", pong)
+	}
+	return nil
+}
+
+func (c *redisBasicClient) MSet(ctx context.Context, keys []string, values [][]byte) error {
+	var cancel context.CancelFunc
+	if c.timeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, c.timeout)
+		defer cancel()
+	}
+
+	pipe := c.rdb.TxPipeline()
+	for i := range keys {
+		pipe.Set(ctx, keys[i], values[i], c.expiration)
+	}
+	_, err := pipe.Exec(ctx)
+	return err
+}
+
+func (c *redisBasicClient) MGet(ctx context.Context, keys []string) ([][]byte, error) {
+	var cancel context.CancelFunc
+	if c.timeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, c.timeout)
+		defer cancel()
+	}
+
+	cmd := c.rdb.MGet(ctx, keys...)
+	if err := cmd.Err(); err != nil {
+		return nil, err
+	}
+
+	ret := make([][]byte, len(keys))
+	for i, val := range cmd.Val() {
+		if val != nil {
+			ret[i] = StringToBytes(val.(string))
+		}
+	}
+	return ret, nil
+}
+
+func (c *redisBasicClient) Close() error {
+	return c.rdb.Close()
+}
+
+type redisSentinelClient struct {
+	logger     log.Logger
+	expiration time.Duration
+	timeout    time.Duration
+	rdb        *redis.SentinelClient
+	opt        *redis.Options
+}
+
+func newRedisSentinelClient(cfg *RedisConfig, logger log.Logger) *redisSentinelClient {
+	opt := &redis.Options{
+		Addr:        cfg.Endpoint,
+		Password:    cfg.Password.Value,
+		PoolSize:    cfg.PoolSize,
+		IdleTimeout: cfg.IdleTimeout,
+		MaxConnAge:  cfg.MaxConnAge,
+	}
+	if cfg.EnableTLS {
+		opt.TLSConfig = &tls.Config{}
+	}
+	return &redisSentinelClient{
+		logger:     logger,
+		expiration: cfg.Expiration,
+		timeout:    cfg.Timeout,
+		rdb:        redis.NewSentinelClient(opt),
+		opt:        opt,
+	}
+}
+
+func (c *redisSentinelClient) getMaster(ctx context.Context) (*redisBasicClient, error) {
```

Comment:
We create a new TCP connection for each `MSet()` and `MGet()`. I believe this is not feasible. I think what we want is keeping a pool of masters (like a `map`) and reuse a master client from the pool if already exists. I'm OK calling `c.rdb.Masters()` for every single `MSet()` and `MGet()` (if that's the way Redis sentinel is expected to work) but then we should reuse master clients/connections whenever we get back the same Redis master.

Reply:
Thanks @pracucci , I'm already looking into this to find the right approach. I will update the PR shortly.

Reply:
I checked the latest changes and I'm not sure how it's different then the previous version. You're still creating a new  `redisBasicClient` for each `MSet()` and `MGet()` operation, which means creating a new TCP connection for each operation, which I believe is not feasible.

What I think you should do is:
1. Call `c.rdb.GetMasterAddrByName()` to get the master address
2. Check if the client is in the local pool of already create clients
  - If yes, use it
  - If not, create a new `redisBasicClient`, store it in the pool and return it